src/tasks/DockerBuildTaskProvider.ts

```diff
@@ -25,59 +25,38 @@ export interface DockerBuildTask extends Task {
-    }
-
-    public resolveTask(task: DockerBuildTask, token?: CancellationToken): ProviderResult<Task> {
-        return callWithTelemetryAndErrorHandling(
-            'docker-build-resolve',
-            async (actionContext: IActionContext) => {
-                const taskPlatform = getPlatform(task.definition);
-                actionContext.telemetry.properties.platform = taskPlatform;
-
-                return await this.resolveTaskInternal(
-                    {
-                        folder: task.scope as WorkspaceFolder,
-                        platform: taskPlatform,
-                        actionContext: actionContext,
-                        cancellationToken: token,
-                    },
-                    task
-                );
-            }
-        );
-    }
+export class DockerBuildTaskProvider extends DockerTaskProviderBase {
+    constructor(helpers: { [key in DockerPlatform]: TaskHelper }) { super('docker-build', helpers) }
 
     // TODO: Skip if image is freshly built
-    private async resolveTaskInternal(context: DockerBuildTaskContext, task: DockerBuildTask): Promise<Task> {
+    protected async executeTaskInternal(context: DockerBuildTaskContext, task: DockerBuildTask): Promise<void> {
         const definition = cloneObject(task.definition);
         definition.dockerBuild = definition.dockerBuild || {};
 
-        if (!context.folder) {
-            throw new Error(`Unable to determine task scope to execute docker-build task '${task.name}'.`);
-        }
-
         const helper = this.getHelper(context.platform);
 
-        if (helper) {
-            definition.dockerBuild = await helper.resolveDockerBuildOptions(context, definition);
+        if (helper && helper.preBuild) {
+            await helper.preBuild(context, definition);
+            throwIfCancellationRequested(context);
         }
 
-        await this.validateResolvedDefinition(context, definition.dockerBuild);
+        if (helper) {
+            definition.dockerBuild = await helper.getDockerBuildOptions(context, definition);
+            await this.validateResolvedDefinition(context, definition.dockerBuild);
+            throwIfCancellationRequested(context);
+        }
 
         const commandLine = await this.resolveCommandLine(definition.dockerBuild);
-        // TODO : addDockerSettingsToEnv?
-        return new Task(
-            task.definition,
-            task.scope,
-            task.name,
-            task.source,
-            new ShellExecution(commandLine[0], commandLine.slice(1)),
-            task.problemMatchers);
+
+        // Because BuildKit outputs everything to stderr, we will not treat output there as a failure
+        await context.shell.executeCommandInTerminal(commandLine, context.folder, /* rejectOnStdError: */ false, context.cancellationToken);
+        throwIfCancellationRequested(context);
+
+        context.imageName = definition.dockerBuild.tag;

```

Comment:
This is intended to set the default `imageName`?  Might `imageName` already be set at this point (or maybe could be in the future)?  Maybe only set if not already set?

Reply:
The goal was to attach the image name to `context` before it gets passed to the post-build tasks. I suppose it's redundant with the config; we could just pass that.

The analogous is to attach the container ID to the context before it gets passed to the post-run tasks. Container ID is not redundant with config though since it isn't known until the container is started.